src/download.rs

```diff
@@ -0,0 +1,300 @@
+//! # Download large messages manually.
+
+use anyhow::{anyhow, Result};
+use deltachat_derive::{FromSql, ToSql};
+use serde::{Deserialize, Serialize};
+
+use crate::config::Config;
+use crate::context::Context;
+use crate::imap::{Imap, ImapActionResult};
+use crate::job::{self, Action, Job, Status};
+use crate::message::{Message, MsgId};
+use crate::param::Params;
+use crate::{job_try, EventType};
+use std::cmp::max;
+
+/// Download limits should not be used below `MIN_DOWNLOAD_LIMIT`.
+///
+/// Some messages as non-delivery-reports (NDN) or read-receipts (MDN)
+/// need to be downloaded completely to handle them correctly,
+/// eg. to assign them to the correct chat.
+/// As these messages are typically small,
+/// they're catched by `MIN_DOWNLOAD_LIMIT`.
+const MIN_DOWNLOAD_LIMIT: u32 = 32768;
+
+/// If `delete_server_after` is set to small timeouts (eg. "at once"),
+/// the user might have no chance to actually download a message.
+/// `MIN_DELETE_SERVER_AFTER` increases the timeout in this case.
+pub const MIN_DELETE_SERVER_AFTER: i64 = 48 * 60 * 60;
+
+#[derive(
+    Debug,
+    Display,
+    Clone,
+    Copy,
+    PartialEq,
+    Eq,
+    FromPrimitive,
+    ToPrimitive,
+    FromSql,
+    ToSql,
+    Serialize,
+    Deserialize,
+)]
+#[repr(u32)]
+pub enum DownloadState {
+    Done = 0,
+    Available = 10,
+    Failure = 20,
+    InProgress = 1000,
+}
+
+impl Default for DownloadState {
+    fn default() -> Self {
+        DownloadState::Done
+    }
+}
+
+impl Context {
+    // Gets validated download limit or 0 for "no limit".
+    pub(crate) async fn get_download_limit(&self) -> Result<u32> {
+        let download_limit = self.get_config_int(Config::DownloadLimit).await?;
+        if download_limit <= 0 {
+            Ok(0)
+        } else {
+            Ok(max(MIN_DOWNLOAD_LIMIT, download_limit as u32))
+        }
+    }
+}
+
+impl MsgId {
+    /// Adds the job `Action::DownloadMsg` to download a message.
+    pub async fn download_full(self, context: &Context) -> Result<()> {
+        let msg = Message::load_from_db(context, self).await?;
+        match msg.get_download_state() {
+            DownloadState::Done => return Err(anyhow!("Nothing to download.")),
+            DownloadState::InProgress => return Err(anyhow!("Download already in progress.")),
```

Comment:
What if the message is stuck in this state and the job is removed due to retry limit or scheduled for several hours later due to backoff, or not added at all because `job::add` failed?

Not sure how to make it robust properly, I'd probably avoid using jobs at all for this task and simply download all "in progress" messages at the start of the message download loop. This way "in progress" state will be fully determined by `msgs` table.

Reply:
> What if the message is stuck in this state and the job is removed due to retry limit or scheduled for several hours later due to backoff

a download job is tried only once, if that fails for whatever reason, the job is removed and the user has to hit "Download" again.

reason is to not waste traffic on retries when eg. the network is not "good enough" or the download fails because of server failures always at 99% - in all these cases, the user has to hit "Download" again to try over.

so, while of course things can be improved (also a "Cancel download" would be nice at some point), i think, the overall state is not so bad already.